File: Find Duplicate Subtrees.java
class Solution {
    Map<String, Integer> map;
    List<TreeNode> res;
    
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        map = new HashMap<>();
        res = new ArrayList<>();
        dfs(root);
        return res;
    }
    
    private String dfs(TreeNode node) {
        if (node == null) {
            return "#";
        }
        String serial = node.val + "," + dfs(node.left) + "," + dfs(node.right);
        map.put(serial, map.getOrDefault(serial, 0) + 1);
        if (map.get(serial) == 2) {
            res.add(node);
        }
        return serial;
    }
}

File: Find Duplicate Subtrees.cpp
class Solution {
public:

unordered_map<string,int>mp;
vector<TreeNode*>v;
string solve(TreeNode* root){

if(root==NULL)return "$";
string s="";


 s=s+to_string(root->val);
 s=s+"#"+solve(root->left);
 s=s+"#"+solve(root->right);

mp[s]++;

if(mp[s]==2){
   v.push_back(root);
}
return s;

}

vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
solve(root);
return v;

}
};
